### PR TITLE
Make panels consistent

### PR DESCRIPTION
**Pull request by predragnikolic**
This is a draft that aims to make the diagnostics, references and rename panel look the same.

Diagnostics before:
Screenshot 2020-12-12 at 16 21 48

Diagnostics after:
Screenshot 2020-12-12 at 16 23 11

References before:
Screenshot 2020-12-12 at 16 21 58

References after:
Screenshot 2020-12-12 at 16 23 18

Rename before:
Screenshot 2020-12-12 at 16 22 05

Rename after:
Screenshot 2020-12-12 at 16 23 25

##### Questions

1. Should we have a \n after each as a separator between filenames?
```
plugin/symbols.py:
	   1:28   from .core.protocol import Request, Range
plugin/core/protocol.py:
	  98:7    class Request:
```
or
```
plugin/symbols.py:
	   1:28   from .core.protocol import Request, Range

plugin/core/protocol.py:
	  98:7    class Request:
```

2. I think I can reuse the same syntax file for references to both references and rename panels? Should I delete the `Rename.sublime-syntax` file? Please give me a suggestion on what to do here. :)

3. The line_regex now must contain a space in front. 
I put a `\t` in front of any line diagnostics, is that ok change?

### CODE REVIEW

**Comment by rwols** [comment]
Looks like a good opportunity to refactor the "rename" and "references" panel to a common builder class. We're going to need such a class anyway, as we'll also be needing to present the "callees" and "callers" from the new set of call hierarchy requests of 3.16: https://github.com/sublimelsp/LSP/issues/1429

**Comment by predragnikolic** [comment]
> a common builder class 

Something like this? 
```python
class Panel:
    def __init__(self, window: sublime.Window, name: str, unlisted=False):
        """
        :param      name:      The name of the output panel
        :param      unlisted:  Control if the output panel should be listed in the panel switcher
        """
        self.window = window
        self.panel = self.window.create_output_panel(name)
        self.name = "output.{}".format(name)

    def is_open(self):
        return self.window.active_panel() == self.name

    def open(self):
        self.window.run_command("show_panel", {"panel": self.name})

    def hide(self):
        self.window.run_command("hide_panel", {"panel": self.name})

    def update(self, content: str):
        self.panel.run_command("lsp_update_panel", {"characters": content})

    def destroy(self):
        self.window.destroy_output_panel(self.name)
```

Not sure what you have envisioned :)

File: plugin/core/views.py [hunk 1]
```diff
@@ -564,7 +564,7 @@ def format_diagnostic_for_panel(diagnostic: Diagnostic) -> Tuple[str, Optional[i
             formatted.append(code)
     lines = diagnostic.message.splitlines() or [""]
     # \u200B is the zero-width space
-    result = "{:>4}:{:<4}{:<8}{} \u200B{}".format(
+    result = "\t{:>4}:{:<4}{:<8}{} \u200B{}\n".format(
```
Comment (rwols) [inline plugin/core/views.py:567]: The `\n` doesn't work for multi-line messages. You'd need to append the additional `\n` somewhere at the end of this function, after the additional indented lines of the message have been added.

Having said that, I prefer no empty lines :)

Comment (rwols) [reply plugin/core/views.py:567]: Actually, it's better if the WindowManager took care of that extra `\n`, if we're going to add that. But again I prefer no `\n` :)

File: plugin/core/windows.py [hunk 2]
```diff
@@ -410,7 +410,7 @@ def handle_show_message(self, session: Session, params: Any) -> None:
         sublime.status_message("{}: {}".format(session.config.name, extract_message(params)))
 
     def update_diagnostics_panel_async(self) -> None:
-        to_render = []  # type: List[str]
+        to_render = ""  # type: str
```
Comment (rwols) [inline plugin/core/windows.py:413]: What makes you think appending strings is faster than joining a list?

Comment (predragnikolic) [reply plugin/core/windows.py:413]: I haven't tought about what is faster.
I saw the way the content for rename and reference was created, 
so I wanted to to use string concatenation to make it consistent for all three:

https://github.com/sublimelsp/LSP/blob/st4000-exploration/plugin/references.py#L123
https://github.com/sublimelsp/LSP/blob/st4000-exploration/plugin/rename.py#L163

That is the reason why I did that. :)

**Comment by rwols** [comment]
> Not sure what you have envisioned :)

A builder class... to abstract away all of the boring indent logic and formatting logic. So something like this:


```python
class PanelBuilder:

    def preamble(self, text: str) -> None:
        """Set the preamble text"""
        ...

    def filename(self, filename: str) -> None:
        """Render a filename to the panel"""
        ...

    def diagnostic(self, diagnostic: Diagnostic) -> None:
        """Render a diagnostic to the panel"""
        ...

    def location(self, range: Dict[str, Any], identifier: Optional[str] = None) -> None:
        """Render a reference/rename location"""
        ...

    def __str__(self) -> str:
        """Flush the content as a string"""
        ...
```

does that make sense?

File: plugin/core/windows.py (hunk 2 above)
Comment (rwols) [reply plugin/core/windows.py:413]: `''.join` is the preferred way to build up chunks of a larger string: https://stackoverflow.com/a/1350289

**Comment by predragnikolic** [comment]
@rwols I know I haven't implemented your suggestion yet. (for the PanelBuilder)
but I wanted to build a `Panel` class that abstracts away the interactions with the panel api.

My question is, should I revert the 7901a7c? 
If the abstraction is too much, or not necessary at the moment I can revert it with no problem :)

**Comment by predragnikolic** [comment]
I reverted the changes 2a962e0 
Can PanelBuilder also be implemented in a different PR?
I just wanted the panels to be consistent :)

File: plugin/core/windows.py (hunk 2 above)
Comment (predragnikolic) [reply plugin/core/windows.py:413]: List or string?
Something tells me that it should be a list :)

**Comment by rwols** [comment]
## 1

> Should we have a \n after each as a separator between filenames?

I don't see the need for it to be honest.

## 2

> Should I delete the Rename.sublime-syntax file?

Sounds good.

## 3

> The line_regex now must contain a space in front. I put a \t in front of any line diagnostics, is that ok change?

I don't understand why this is needed? Why doesn't SublimeLinter use a tab? http://www.sublimelinter.com/en/stable/

## 4

Please don't use string concatenation in windows.py

File: plugin/core/windows.py (hunk 2 above)
Comment (predragnikolic) [reply plugin/core/windows.py]: done with 02d6859

File: plugin/core/views.py (hunk 1 above)
Comment (predragnikolic) [reply plugin/core/views.py]: I removed the `\n` and `\t` from those lines with 02d6859 
```diff
- \t{:>4}:{:<4}{:<8}{} \u200B{}\n\
+ {:>5}:{:<4}{:<8}{} \u200B{}\
```

**Comment by predragnikolic** [comment]
4 - is done.
 3 - is done.
 2 - is done.
 1 - Although a `\n` is not necessary I would argue that a new line gives a visual separation of the results which help readability. Because of that I would keep the \n between filenames. :)

File: plugin/core/windows.py [hunk 3]
```diff
@@ -439,6 +439,7 @@ def update_diagnostics_panel_async(self) -> None:
                 if offset is not None and code is not None and href is not None:
                     prephantoms.append((row, offset, code, href))
                 row += content.count("\n") + 1
+            to_render.append("")  # add spacing between filenames
```
Comment (rwols) [inline plugin/core/windows.py:442]: Does this require another `row += 1` ?

Comment (rwols) [reply plugin/core/windows.py:442]: You can check this by running with e.g. eslint and creating diagnostics in at least two files.

Comment (predragnikolic) [reply plugin/core/windows.py:442]:
```diff
 to_render.append("")  # add spacing between filenames
+ row += 1
```
I don't see a difference.
with `row += 1`:

with
or without `row += 1`:
without

But I have noticed on the first line that `typescript:6133` is not highlighted. (I am not a ST syntax expert, not sure how to fix it)
```
     1:1   hint    'styles' is declared but its value is never read. ​typescript:6133
````

Comment (rwols) [reply plugin/core/windows.py:442]: I'm talking about the clickable links, make sure to check those:

image

Comment (predragnikolic) [reply plugin/core/windows.py:442]: Thanks for pointing this out.

> Does this require another row += 1 ?

It turns out that it does: 

Screenshot from 2020-12-18 21-21-23
the phantom placement is messed up.

It is fixed with 0bcd931

Comment (rchl) [reply plugin/core/windows.py:442]: > But I have noticed on the first line that `typescript:6133` is not highlighted. (I am not a ST syntax expert, not sure how to fix it)

There is missing `pop: true` in here: https://github.com/sublimelsp/LSP/blob/0bcd93110d98f74879f274b2cb10fa50aec6a663/Syntaxes/Diagnostics.sublime-syntax#L54-L55

Comment (predragnikolic) [reply plugin/core/windows.py:442]: Thanks @rchl,
I included the fix e729f45

**Review by rwols — APPROVED** [review]

**Pull request merged**